Mage/src/main/java/mage/abilities/effects/common/combat/RedirectAttackerTargetEffect.java

```diff
@@ -0,0 +1,104 @@
+package mage.abilities.effects.common.combat;
+
+import java.util.LinkedHashSet;
+import java.util.Set;
+import java.util.UUID;
+
+import mage.abilities.Ability;
+import mage.abilities.effects.OneShotEffect;
+import mage.constants.CardType;
+import mage.constants.Outcome;
+import mage.filter.FilterPermanent;
+import mage.filter.StaticFilters;
+import mage.filter.predicate.Predicates;
+import mage.game.Game;
+import mage.game.combat.CombatGroup;
+import mage.game.permanent.Permanent;
+import mage.players.Player;
+import mage.target.common.TargetDefender;
+
+public class RedirectAttackerTargetEffect extends OneShotEffect {
+    
+    private final boolean includePermanents;
+    private static final FilterPermanent filter = new FilterPermanent("permanent");
+
+    static {
+        filter.add(Predicates.or(
+            CardType.PLANESWALKER.getPredicate(), 
+            CardType.BATTLE.getPredicate()
+        ));
+    }
+
+    public RedirectAttackerTargetEffect(boolean includePermanents) {
+        super(Outcome.Benefit);
+        this.includePermanents = includePermanents;
+    }
+
+    public RedirectAttackerTargetEffect(final RedirectAttackerTargetEffect effect) {
+        super(effect);
+        this.includePermanents = effect.includePermanents;
+    }
+
+    @Override
+    public RedirectAttackerTargetEffect copy() {
+        return new RedirectAttackerTargetEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player controller = game.getPlayer(source.getControllerId());
+        if (controller == null) {
+            return false;
+        }
+        
+        // Get combat group for attacking creatures.
+        for (UUID id : getTargetPointer().getTargets(game, source)) {
+
+            Permanent attackingCreature = game.getPermanent(id);
+            if (attackingCreature == null) {
+                continue;
+            }
+
+            CombatGroup combatGroupTarget = game.getCombat().findGroup(attackingCreature.getId());
+            if (combatGroupTarget == null) {
+                continue;
+            }
+
+            // Reselecting which player or permanent a creature is attacking ignores all requirements, restrictions, and costs associated with attacking.
+
+            // Update possible defender
+            Set<UUID> defenders = new LinkedHashSet<>();
+            for (UUID playerId : game.getCombat().getAttackablePlayers(game)) {
+                defenders.add(playerId);
+                if (includePermanents) {
+                    for (Permanent permanent : game.getBattlefield().getAllActivePermanents(filter, playerId, game)) {
```

Comment:
Does this exclude attacking a planeswalker controlled by attacking player? Or is that handled later by TargetDefender?

Reply:
the only permanents it gets are from the players that are attackable, so I assume that would exclude the attackers permanents

Reply:
Right. Well in that case, does it work correctly with battles?

Reply:
Is there a better Combat function for getting all possible attackable targets?

Reply:
Probably not.
I'd recommend splitting the logic. Add planeswalkers the way you are now (but with existing static filter). Then add battles separately, excluding based on protector. Any existing functionality would be in the battle PR #10156

Reply:
I like the way you're handling it now

Reply:
Yeah, I figured just filtering by controller/protector was the cleaner way of going about it